src/main/java/org/sagebionetworks/bridge/services/Exporter3Service.java

```diff
@@ -0,0 +1,522 @@
+
+        // Finally, enable Exporter 3.0 if it's not already enabled.
+        if (!app.isExporter3Enabled()) {
+            app.setExporter3Enabled(true);
+            isAppModified = true;
+        }
+
+        // Update app if necessary. (Don't need to update any admin fields.)
+        if (isAppModified) {
+            appService.updateApp(app, false);
+        }
+
+        return ex3Config;
+    }
+
+    // Package-scoped for unit tests.
+    String getNameScopingToken() {
+        return SecureTokenGenerator.NAME_SCOPE_INSTANCE.nextToken();
+    }
+
+    /** Complete an upload for Exporter 3.0, and also export that upload. */
+    public void completeUpload(App app, Upload upload) {
+        // Create record.
+        HealthDataRecordEx3 record = HealthDataRecordEx3.createFromUpload(upload);
+        record = healthDataEx3Service.createOrUpdateRecord(record);
+
+        // Export upload. Note that we complete the export synchronously in the request thread, because it doesn't take
+        // that long, and app developers prefer synchronous.
+        exportUpload(app, upload, record);
+    }
+
+    // This is separate because in the future, we might need a separate redrive process in the future.
+    private void exportUpload(App app, Upload upload, HealthDataRecordEx3 record) {
+        // Check that app is configured for export.
+        Exporter3Configuration exporter3Config = app.getExporter3Configuration();
+        if (exporter3Config == null || !exporter3Config.isConfigured()) {
+            // Exporter not enabled. Skip.
+            return;
+        }
+
+        String appId = app.getIdentifier();
+        String uploadId = upload.getUploadId();
+        try {
+            // Set the exportedOn time on the record. This will be propagated to both S3 and Synapse.
+            record.setExportedOn(DateUtils.getCurrentMillisFromEpoch());
+
+            // Copy the file to the raw health data bucket. This includes folderization.
+            ExportContext exportContext;
+            if (upload.isEncrypted()) {
+                exportContext = decryptAndUploadFile(upload, record);
+            } else {
+                exportContext = copyUploadToHealthDataBucket(upload, record);
+            }
+
+            // Upload to Synapse.
+            exportToSynapse(exporter3Config, upload, exportContext);
+
+            // Mark record as exported.
+            record.setExported(true);
+            healthDataEx3Service.createOrUpdateRecord(record);
+        } catch (Exception ex) {
+            LOG.error("Exporter 3 error processing upload, app=" + appId + ", upload=" + uploadId + ": " +
+                    ex.getMessage(), ex);
+
+            // Propagate 4XX erors.
+            if (ex instanceof BridgeServiceException) {
+                int status = ((BridgeServiceException) ex).getStatusCode();
+                if (status >= 400 && status < 500) {
+                    throw new BadRequestException(ex);
+                }
+            }
```

Comment:
Error propagation may not be necessary anymore now that there's no cleanup step.